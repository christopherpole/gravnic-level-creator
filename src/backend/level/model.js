var mongoose = require('mongoose');

//  Level schema
var LevelSchema = new mongoose.Schema({
  name: { type: String, required: true },
  tiles: { type: Array, required: true },
});

module.exports = mongoose.model('Level', LevelSchema);
