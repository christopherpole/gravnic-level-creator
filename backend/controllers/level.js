const express = require('express');
const bodyParser = require('body-parser');
const winston = require('winston');
const Level = require('../models/level');

const router = express.Router();
router.use(bodyParser.urlencoded({ extended: true }));
router.use(bodyParser.json());

router.post('/', (req, res) => {
  Level.create(
    {
      name: req.body.name,
      tiles: req.body.tiles,
    },
    (err, level) => {
      if (err) {
        winston.error(err.message);

        return res.status(500).send(err.message);
      }

      return res.status(201).send(level);
    },
  );
});

router.get('/', (req, res) => {
  Level.find({}, (err, levels) => {
    if (err) {
      winston.error(err.message);

      return res.status(500).send(err.message);
    }

    return res.status(200).send(levels);
  });
});

router.get('/:levelId', (req, res) => {
  Level.findById(req.params.levelId, (err, level) => {
    if (err) {
      winston.error(err.message);

      return res.status(500).send(err.message);
    }

    if (!level) {
      return res.status(404).send();
    }

    return res.status(200).json(level);
  });
});

//  Technically a PATCH but whatever
router.put('/:levelId', (req, res) => {
  Level.findByIdAndUpdate(
    req.params.levelId,
    req.body,
    { upsert: false, new: true },
    (err, level) => {
      if (err) {
        winston.error(err.message);

        return res.status(500).send(err.message);
      }

      if (!level) {
        return res.status(404).send();
      }

      return res.status(200).json(level);
    },
  );
});

router.delete('/:levelId', (req, res) => {
  Level.findByIdAndRemove(req.params.levelId, (err, level) => {
    if (err) {
      winston.error(err.message);

      return res.status(500).send(err.message);
    }

    if (!level) {
      return res.status(404).send();
    }

    return res.status(204).send();
  });
});

module.exports = router;
