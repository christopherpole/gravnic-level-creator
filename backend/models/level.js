const mongoose = require('mongoose');
const winston = require('winston');
mongoose.Promise = require('bluebird');

//  Tile schema
const TileSchema = new mongoose.Schema({
  position: { type: Number, required: true },
  selectedTileId: { type: Number, required: true },
});

TileSchema.set('toJSON', {
  transform: (doc, ret) => {
    const rtn = { ...ret };
    delete rtn._id;
    return rtn;
  },
});

//  Level schema
const LevelSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  tiles: {
    type: [TileSchema],
    required: true,
  },
});

LevelSchema.post('save', level => {
  winston.info(`Created level with the id of ${level._id}`);
});

LevelSchema.post('remove', level => {
  winston.info(`Removed level with the id of ${level._id}`);
});

LevelSchema.set('toJSON', {
  transform: (doc, ret) => {
    const rtn = { ...ret };

    rtn.id = rtn._id;
    delete rtn._id;
    delete rtn.__v;

    return rtn;
  },
});

module.exports = mongoose.model('Level', LevelSchema);
