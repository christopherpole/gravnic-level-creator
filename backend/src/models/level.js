const mongoose = require('mongoose');
const winston = require('winston');
mongoose.Promise = require('bluebird');

//  Tile schema
const TileSchema = new mongoose.Schema({
  position: { type: Number, required: true },
  selectedTileId: { type: String, required: true },
});

TileSchema.set('toJSON', {
  transform: (doc, ret) => {
    const rtn = { ...ret };
    delete rtn._id;
    return rtn;
  },
});

//  Level schema
const LevelSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  tiles: {
    type: [TileSchema],
    required: true,
  },
  stars: {
    type: [Number],
    required: true,
  },
  position: {
    type: Number,
    required: true,
  },
  solution: {
    type: mongoose.Schema.Types.Mixed,
    default: undefined,
  },
  maxMoves: {
    type: Number,
    default: undefined,
  },
});

LevelSchema.post('save', level => {
  winston.info(`Created level with the id of ${level._id}`);
});

LevelSchema.post('remove', level => {
  winston.info(`Removed level with the id of ${level._id}`);
});

LevelSchema.set('toJSON', {
  transform: (doc, ret) => {
    const rtn = { ...ret };

    rtn.id = rtn._id;
    delete rtn._id;
    delete rtn.__v;

    return rtn;
  },
});

module.exports = mongoose.model('Level', LevelSchema);
