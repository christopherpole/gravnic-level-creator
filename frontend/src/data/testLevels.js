import { MOVE_DOWN, MOVE_UP } from 'gravnic-game';

export default [
  {
    id: '1',
    name: 'Test level 1',
    solution: [MOVE_DOWN, MOVE_UP],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '2',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '2',
      },
      {
        position: 48,
        selectedTileId: '4',
      },
      {
        position: 49,
        selectedTileId: '1',
      },
      {
        position: 50,
        selectedTileId: '1',
      },
      {
        position: 51,
        selectedTileId: '1',
      },
      {
        position: 52,
        selectedTileId: '2',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 1,
  },
  {
    id: '2',
    name: 'Test level 2',
    solution: [MOVE_UP],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '2',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '4',
      },
      {
        position: 48,
        selectedTileId: '2',
      },
      {
        position: 49,
        selectedTileId: '1',
      },
      {
        position: 50,
        selectedTileId: '2',
      },
      {
        position: 51,
        selectedTileId: '2',
      },
      {
        position: 52,
        selectedTileId: '2',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 2,
  },
  {
    id: '3',
    name: 'Test level 3',
    solution: [],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '4',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '2',
      },
      {
        position: 48,
        selectedTileId: '2',
      },
      {
        position: 49,
        selectedTileId: '2',
      },
      {
        position: 50,
        selectedTileId: '2',
      },
      {
        position: 51,
        selectedTileId: '1',
      },
      {
        position: 52,
        selectedTileId: '1',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 3,
  },
];
