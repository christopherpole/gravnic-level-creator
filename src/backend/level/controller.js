const express = require('express');
const bodyParser = require('body-parser');
const Level = require('./model');

const router = express.Router();
router.use(bodyParser.urlencoded({ extended: true }));
router.use(bodyParser.json());

router.post('/', (req, res) => {
  Level.create({
    name: req.body.name,
    tiles: req.body.tiles,
  }, (err, level) => {
    if (err) {
      return res
        .status(500)
        .send('There was a problem adding the information to the database.');
    }

    return res.status(200).send(level);
  });
});

router.get('/', (req, res) => {
  Level.find({}, (err, levels) => {
    if (err) {
      return res
        .status(500)
        .send('There was a problem finding the levels.');
    }

    return res.status(200).send(levels);
  });
});

module.exports = router;
