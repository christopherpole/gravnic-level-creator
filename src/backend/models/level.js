const mongoose = require('mongoose');
const winston = require('winston');
mongoose.Promise = require('bluebird');

//  Level schema
const LevelSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  tiles: {
    type: [{
      position: { type: Number, required: true },
      selectedTileId: { type: Number, required: true },
    }],
    required: true,
  },
});

LevelSchema.post('save', (level) => {
  winston.info(`Created level with the id of ${level._id}`);
});

LevelSchema.post('remove', (level) => {
  winston.info(`Removed level with the id of ${level._id}`);
});

module.exports = mongoose.model('Level', LevelSchema);
