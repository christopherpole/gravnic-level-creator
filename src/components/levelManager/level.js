import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { selectLevel } from '../../actions/levelManager';

export const Wrapper = styled.li`
  border: 2px solid white;
  cursor: pointer;
  list-style-type: none;
  margin-bottom: ${props => props.theme.structureSpacing};
  padding: ${props => props.theme.structureSpacing};

  &:last-child {
    margin-bottom: 0;
  }

  ${props =>
    props.isSelected &&
    css`
      border-color: yellow;
    `};
`;

export const Level = ({ id, name, isSelected, selectLevelAction }) => (
  <Wrapper
    onClick={() => {
      selectLevelAction(id);
    }}
    isSelected={isSelected}
  >
    <p>{name}</p>
  </Wrapper>
);

Level.defaultProps = {
  isSelected: false,
};

Level.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  isSelected: PropTypes.bool,
  selectLevelAction: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  selectLevelAction: bindActionCreators(selectLevel, dispatch),
});

export default connect(null, mapDispatchToProps)(Level);
