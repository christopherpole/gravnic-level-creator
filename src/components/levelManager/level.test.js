import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { spy } from 'sinon';

import { Level, Wrapper } from './level';
import testLevels from '../../data/testLevels';

configure({ adapter: new Adapter() });

describe('The level', () => {
  let props;

  beforeEach(() => {
    props = {
      ...testLevels[0],
      selectLevelAction: () => {},
    };
  });

  it('Renders without exploding', () => {
    const level = shallow(<Level {...props} />);

    expect(level).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const level = shallow(<Level {...props} />);

    expect(toJson(level)).toMatchSnapshot();
  });

  it('Fires the select level action when clicking on a level', () => {
    const levelClickSpy = spy();
    const level = shallow(<Level {...props} selectLevelAction={levelClickSpy} />);
    const levelWrapper = level.find(Wrapper);
    levelWrapper.simulate('click');

    expect(levelClickSpy.calledOnce).toBe(true);
    expect(levelClickSpy.calledWith(testLevels[0].id)).toBe(true);
  });
});
