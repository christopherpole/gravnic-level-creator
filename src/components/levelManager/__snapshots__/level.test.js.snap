// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level Matches the current snapshot 1`] = `
<styled.li
  isSelected={false}
  onClick={[Function]}
>
  <p>
    Test level
  </p>
</styled.li>
`;
